contrib/platform/src/com/sun/jna/platform/win32/Kernel32.java

```diff
@@ -4178,4 +4178,101 @@ Pointer VirtualAllocEx(HANDLE hProcess, Pointer lpAddress, SIZE_T dwSize,
+     *                       characters. Do not include the name of the
+     *                       executable in the command line; this function adds
+     *                       it for you.
+     *
+     * <p>If this parameter is NULL or an empty string, the previously registered
+     * command line is removed. If the argument contains spaces, use quotes
+     * around the argument.</p>
+     *
+     * @param dwFlags        This parameter can be 0 or one or more of the
+     *                       following values.
+     *
+     * <table>
+     * <tr>
+     * <th>Value</th><th>Meaning</th>
+     * </tr>
+     * <tr><td>RESTART_NO_CRASH (1)</td><td>Do not restart the process if it terminates due to an unhandled exception.</td></tr>
+     * <tr><td>RESTART_NO_HANG (2)</td><td>Do not restart the process if it terminates due to the application not responding.</td></tr>
+     * <tr><td>RESTART_NO_PATCH (4)</td><td>Do not restart the process if it terminates due to the installation of an update.</td></tr>
+     * <tr><td>RESTART_NO_REBOOT (8)</td><td>Do not restart the process if the computer is restarted as the result of an update. </td></tr>
+     * </table>
+     *
+     * @return This function returns S_OK on success or one of the following
+     *         error codes.
+     * <table>
+     * <tr><th>Return code</th><th>Description</th></tr>
+     * <tr><td>E_FAIL</td><td>Internal error.</td></tr>
+     * <tr><td>E_INVALIDARG</td><td>The specified command line is too
+     * long.</td></tr>
+     * </table>
+     */
+    HRESULT RegisterApplicationRestart(char[] pwzCommandline, int dwFlags);
+
+    /**
+     * Removes the active instance of an application from the restart list.
+     *
+     * @return This function returns S_OK on success or one of the following
+     *         error codes.
+     * <table>
+     * <tr><th>Return code</th><th>Description</th></tr>
+     * <tr><td>E_FAIL</td><td>Internal error.</td></tr>
+     * </table>
+     */
+    HRESULT UnregisterApplicationRestart();
+
+    /**
+     * Retrieves the restart information registered for the specified process.
+     *
+     * @param hProcess       A handle to the process. This handle must have the
+     *                       PROCESS_VM_READ access right.
+     * @param pwzCommandline A pointer to a buffer that receives the restart
+     *                       command line specified by the application when it
+     *                       called the RegisterApplicationRestart function. The
+     *                       maximum size of the command line, in characters, is
+     *                       RESTART_MAX_CMD_LINE. Can be NULL if pcchSize is
+     *                       zero.
+     * @param pcchSize       On input, specifies the size of the pwzCommandLine
+     *                       buffer, in characters.
+     *
+     * <p>
+     * If the buffer is not large enough to receive the command line, the
```

Comment:
Is it ok to copy/paste the Microsoft documentation verbatim? I haven't checked thoroughly if they publish it under a license that allows it.

Reply:
My take on this is, that the documentation must be part of the API and I assume API is not copyrightable/this falls under the fair use assumption. I'm not a lawyer and I'm aware, that this is still under discussion in the US.

Reply:
All the documentation on `docs.microsoft.com` is [open-sourced here](https://github.com/MicrosoftDocs) and [licensed under CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/).  
> You are free to:
> - **Share** — copy and redistribute the material in any medium or format
> - **Adapt** — remix, transform, and build upon the material
> - for any purpose, even commercially.

With only the restriction:
> **Attribution** — You must give appropriate credit, provide a link to the license, and indicate if changes were made. You may do so in any reasonable manner, but not in any way that suggests the licensor endorses you or your use.

While many of the citations from Microsoft docs (formerly MSDN) were specifically linked in the javadocs (with `@see`) it's not universal. To make broad API citations more clear, we could add some sort of document in the project root more clearly documenting this usage "in any reasonable manner" (e.g., ASF projects use a NOTICE file for this purpose).  The attribution could be an [easy one-liner](https://creativecommons.org/use-remix/attribution/).

Reply:
JNA is not ALv2, it is dual-licensed ALv2+LGPLv2. I don't see a problem: Either the API is copyrightable and this is not fair use, then we can pretty much delete the whole platform project, or it is not copyrightable or fair use, in that case it can stay as is.